README.md
# jointSeg

This package implements functions to quickly segment multivariate signals into piecewise constant profiles and a framework to generate realistic copy-number profiles. A typical application is the joint segmentation of total DNA copy numbers and allelic ratios obtained from Single Nucleotide Polymorphism (SNP) microarrays in cancer studies.

# Installation

The package can be installed from github:

# install.packages("devtools")
devtools::install_github("mpierrejean/jointSeg")

# Examples and descriptions
A detailled example can be find in PSSeg.pdf and dataGeneration.pdf


NEWS
Package: jointSeg
=================

Version: 0.7.0 [2015-07-08]
o In 'PSSeg': added argument 'rankTransform' to allow rank-based
segmentation to be performed.

Version: 0.6.4 [2014-12-19]
o Bug fixed for vignette (remove GFLars segmentation)

Version: 0.6.3 [2014-10-27]
o Bugs fixed to submission to CRAN

Version: 0.6.2 [2014-09-09]
o Updates to vignette.
o Updates to CITATION.

Version: 0.6.1 [2014-07-16]
o In 'estimateSd': SPEED UP for method "Hall": vectorization.
o In 'jointSeg', 'PSseg': it is now possible to run 'modelSelection'
 on initial segmentation.  Default behavior of the functions is
 unchanged.
o In 'getCopyNumberDataByResampling', if 'regAnnot' is NULL (the
default), frequencies of regions (0,1), (0,2), (1,1) and (1,2) (the
most common alterations) are set to represent 90% of the regions.

Version: 0.6.0 [2014-06-18]
o Cleanups in DESCRIPTION/NAMESPACE.
o Moved 'doCnaStruct' to 'otherMethods/'.
o Packages passes R CMD check --as-cran.

Version: 0.5.4 [2014-06-06]
o Replaced 'jointSeg' by 'jointseg' in 'inst' scripts.

Version: 0.5.3 [2014-06-04]
o Some changes to the main vignette (now using 'knitr').
o Updates in scripts to perform ASCAT segmentation.

Version: 0.5.2 [2014-05-30]
o Added a vignette to illustrate how to add a new method.

Version: 0.5.1 [2014-05-29]
o Changed Illumina to GSE11976 and Affymetrix to GSE29172
o Bug fixed for PSCBS, DynamicProgramming and CBS methods.
o Updated scripts

Version: 0.5.0 [2014-05-26]
o Major changes in the interface of the main functions 'PSSeg' and
'jointseg'.
o Now 'doGFLars' is a wrapper calling the lower-level 'segmentByGFLars'.
o Now 'doRBS' is a wrapper calling the lower-level 'segmentByRBS'.
o Added 'binMissingValues' so that it may be used not only for GFLars.
o Cleaned up documentation and examples.


Version: 0.4.7 [2014-05-14]
o in 'jointSeg': added argument 'segFUN' and flavor "other" to enable
a user-defined segmentation method to be used.


Version: 0.4.6 [2014-05-09]
o BUG FIX in 'doCBS': replaced start bkp given by CBS by end bkp as we
consider ends of segments for all methods.  Only makes a difference
when there are missing values in the signal.

Version: 0.4.5 [2014-05-06]
o Bug fixed in scripts to build table of the paper.
o Updated mapping for d|het segmentations (median).

Version: 0.4.4 [2014-05-05]
o Added scripts to 'inst/otherMethods' and 'inst/syntheticProfiles'
  to address referees' comments.

Version: 0.4.3 [2014-02-20]
o Added 'inst/figures' to reproduce the figures of the paper.

Version: 0.4.2 [2014-02-14]
o 'plotSeg': Calculation of segment means would fail when a breakpoint
  was detected between the first and second position.

Version: 0.4.1 [2014-02-13]
o Renamed function 'jointSeg' to 'jointseg' to avoid name clash with
  package name.
o Now all documented functions are exported.
o Updates in vignette.

Version: 0.4.0 [2014-02-13]
o Added scripts to reproduce methods comparison of the jointSeg paper
  to dedicated subdirectories.
o Added scripts to reproduce figures of the jointSeg paper to 'figures'.
o Removed defunct package PSCN from 'Suggests' field, and moved function
  'doPSCN' to (newly created) directory 'zzz.defunct'.
o Removed 'R.utils' from 'Depends' field.

Version: 0.3.17 [2014-02-12]
o Updates in 'plotSeg'.

Version: 0.3.16 [2013-12-09]
o 'segmentByNnn' replaced by 'doNnn'.
o Replaced 'segmentByCghseg' by 'doDynamicProgramming'.

Version: 0.3.15 [2013-12-06]
o Wrappers for PSCN and CnaStruct now operate systematically on
  log-scaled copy numbers (ie LRRs).  Thanks to report by Hugues
  Sicotte.
o Segmentation methods can now be run on numeric vectors or
 data.frames (when appropriate).

Version: 0.3.14 [2013-11-29]
o Added flavor 'DP'.

Version: 0.3.13 [2013-11-08]
o Bumped version number to trigger build by R-forge.

Version: 0.3.12 [2013-08-04]
o Added function 'getTpFp' so that performance can be evaluated in a
  scale which does not depend on the signal length.  Updated examples
  and vignette accordingly.

Version: 0.3.11 [2013-05-30]
o Added argument 'regNames' to 'plotSeg' so that region labels are
  plotted if available.

Version: 0.3.10 [2013-05-16]
o Some example code now embedded in 'require()' statements to avoid
  problems in the R CMD check mechanism of R-forge.

Version: 0.3.9 [2013-04-09]
o Added argument 'K' to 'pruneByDP'.

Version: 0.3.8 [2013-03-27]
o Modification in 'segmentByCnaStruct'

Version: 0.3.7 [2013-03-27]
o Added segmentation methods PELT and CnaStruct

Version: 0.3.6 [2013-03-19]
o Added flavor "Hall" to estimate standard deviation in 'estimateSd'.
o Added argument "DP" in 'PSSeg' and 'jointSeg'.
o Added argument "relax" in 'getTprTnr' and definition of true negatives.
o Added return value "rse" in 'segmentByRBS'.
o Added argument "connex" in 'getCopyNumberByResampling'.

Version: 0.3.5 [2013-02-26]
o In 'PSSeg' and 'jointSeg': added argument 'jitter'.
o In getCopyNumberDataByResampling': added argument 'connex' to forces
adjacent regions to be connex.
o In 'getTprTnr': added argument 'relax'.

Version: 0.3.4 [2013-02-18]
o 'segmentByGFLars' now handles missing values.
o In 'PSSeg': Flavor "GFLars" can now be run at full resolution

Version: 0.3.3 [2013-02-18]
o BUG FIX in 'segmentByPSCBS'.
o flavor = 'PSBCS' added.
o Removed segmentByPairedPSCBS.

Version: 0.3.2 [2013-02-14]
o BUG FIX in 'modelSelection': if the dimension is too small, in 'jointSeg' if flavor = 'CBS' or 'PSCN' no using of 'modelSelection'.

Version: 0.3.1 [2013-02-07]
o Bumped acnr dependency (>= 0.1.1) to fix minor case sensitivity issue.
o Reduced graphics size in vignette.

Version: 0.3.0 [2013-01-31]
o Updated vignette.
o Passes R CMD check with no errors.

Version: 0.2.5 [2013-01-31]
o Updated NAMESPACE.
o Moved dilution data sets to new package 'acnr'.

Version: 0.2.4 [2013-01-25]
o Cleanups in doc and return values.
o 'jointSeg' now returns 'bestBkp'
o Removed 'position' from I/O in 'getCopyNumberDataByResampling' and 'PSSeg'.

R/PSSeg.R
PSSeg <- structure(function(#Parent-Specific copy number segmentation
### This function splits (bivariate) copy number signals into
### parent-specific (PS) segments using recursive binary segmentation
                            data,
### Data frame containing the following columns: \describe{
### \item{c:}{Total copy number (logged or non-logged)}
### \item{b:}{Allele B fraction}
### \item{genotype:}{(germline) genotype of the SNP, coded as 0 for AA, 1/2 for AB, 1 for BB}
### }
### These data are assumed to be ordered by genome position.
                            method,## A \code{character} value, the type of segmentation method used. May be one of
### \describe{
###   \item{"RBS"}{Recursive Binary Segmentation, see
### \code{\link{doRBS}}}
###   \item{"GFLars"}{Group fused LARS as described in Bleakley and
###   Vert (2011).}
###   \item{"DP"}{Univariate pruned dynamic programming Rigaill et al (2010) or bivariate dynamic programming}
###   \item{"PSCBS"}{Parent-specific copy number in paired tumor-normal studies using circular binary segmentation by Olshen A. et al
###     (2011)}
###   \item{"other"}{The segmentation method is passed as a function using argument \code{segFUN} (see examples in directory \code{otherMethods}).}
###}
                            stat=NULL,
### A vector containing the names or indices of the columns of \code{Y} to be segmented
                            dropOutliers=TRUE,
### If TRUE, outliers are droped by using DNAcopy package
    rankTransform=FALSE,
### If TRUE, data are replaced by their ranks before segmentation
                            ...,
### Further arguments to be passed to \code{jointSeg}
                            profile=FALSE,
### Trace time and memory usage ?
                            verbose=FALSE
### A \code{logical} value: should extra information be output ? Defaults to \code{FALSE}.
                            ){
    ## Argument
    cn <- colnames(data)
    ecn <- c("c", "b", "genotype") ## expected
    mm <- match(ecn, cn)
    if (any(is.na(mm))) {
        str <- sprintf("('%s')", paste(ecn, collapse="','"))
        stop("Argument 'data' should contain columns named ", str)
    }

    prof <- NULL

    ## - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    ## Pre-processing
    ## - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    n <- nrow(data)
    data$idx <- 1:n
    data$d <- 2*abs(data$b-1/2)
    isHet <- (data[["genotype"]]==0.5)
    data[which(!isHet), "d"] <- NA
    ##details<<Before segmentation, the decrease in heterozygosity
    ##\code{d=2|b-1/2|} defined in Bengtsson et al, 2010 is calculated
    ##from the input data.  \code{d} is only defined for heterozygous
    ##SNPs, that is, SNPs for which \code{data$genotype==1/2}. \code{d}
    ##may be seen as a "mirrored" version of allelic ratios (\code{b}):
    ##it converts them to a piecewise-constant signals by taking
    ##advantage of the bimodality of \code{b} for heterozygous SNPs.
    ##The rationale for this transformation is that allelic ratios
    ##(\code{b}) are only informative for heterozygous SNPs (see
    ##e.g. Staaf et al, 2008).

    if(dropOutliers){
        CNA.object <- CNA(data$c, rep(1, n), 1:n)
        smoothed.CNA.obj <- smooth.CNA(CNA.object)
        data$c <- smoothed.CNA.obj$Sample.1
    }
    ##details<<Before segmentation, the outliers in the copy number
    ##signal are dropped according the method explained by
    ##Venkatraman, E. S. and Olshen, A. B., 2007.

    if (rankTransform) {
        data[, "c"] <- rank(data[, "c"], na="keep")
        data[, "d"] <- rank(data[, "d"], na="keep")
    }

    ## - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    ## Segmentation followed by pruning using dynamic programming
    ## - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

    ##details<<The resulting data are then segmented using the
    ##\code{\link{jointSeg}} function, which combines an initial
    ##segmentation according to argument \code{method} and pruning of
    ##candidate change points by dynamic programming (skipped when the
    ##initial segmentation *is* dynamic programming).
    if (is.null(stat)) {
        ##details<<If argument \code{stat} is not provided, then dynamic
        ##programming is run on the two dimensional statistic
        ##\code{"(c,d)"}.
        mm <- match(c("c", "d"), colnames(data))
        res <- jointSeg(data, method=method, dpStat=mm, ..., profile=profile, verbose=verbose)
    } else {
        ##details<<If argument \code{stat} is provided, then dynamic
        ##programming is run on \code{stat}; in this case we implicitly
        ##assume that \code{stat} is a piecewise-constant signal.
        res <- jointSeg(data, method=method, stat=stat, dpStat=stat, ..., profile=profile, verbose=verbose)
    }
    prof <- rbind(prof, res$prof)



    ##references<<Bengtsson, H., Neuvial, P., & Speed,
    ##T. P. (2010). TumorBoost: Normalization of allele-specific tumor
    ##copy numbers from a single pair of tumor-normal genotyping
    ##microarrays. BMC bioinformatics, 11(1), 245.

    ##references<<Staaf, J., Lindgren, D., Vallon-Christersson, J.,
    ##Isaksson, A., Goransson, H., Juliusson, G., ... & Ringn\'er,
    ##M. (2008). Segmentation-based detection of allelic imbalance and
    ##loss-of-heterozygosity in cancer cells using whole genome SNP
    ##arrays. Genome Biol, 9(9), R136.

    ##seealso<<\code{\link{jointSeg}}
    list(
        bestBkp=res$bestBkp, ##<< Best set of breakpoints after
        ##dynamic programming
        initBkp=res$initBkp, ##<< Results of the initial segmentation,
        ##using 'doNnn', where 'Nnn' corresponds
        ##to argument \code{method}
        dpBkpList=res$dpBkpList, ##<< Results of dynamic programming,
        ##a list of vectors of breakpoint
        ##positions for the best model with k
        ##breakpoints for k=1, 2, ... K where
        ##\code{K=length(initBkp)}
        prof=prof ##<< a \code{matrix} providing time usage (in
        ##seconds) and memory usage (in Mb) for the main
        ##steps of the program.  Only defined if argument
        ##\code{profile} is set to \code{TRUE}
        )
}, ex=function(){
    ## load known real copy number regions
    affyDat <- loadCnRegionData(dataSet="GSE29172", tumorFraction=0.5)

    ## generate a synthetic CN profile
    K <- 10
    len <- 1e4
    sim <- getCopyNumberDataByResampling(len, K, regData=affyDat)
    datS <- sim$profile

    ## run binary segmentation (+ dynamic programming) resRBS <-
    resRBS <- PSSeg(data=datS, method="RBS", stat=c("c", "d"), K=2*K, profile=TRUE)
    resRBS$prof

    getTpFp(resRBS$bestBkp, sim$bkp, tol=5)
    plotSeg(datS, breakpoints=list(sim$bkp, resRBS$bestBkp))
})
############################################################################
## HISTORY:
## 2015-07-08
## o Added argument 'rankTransform' to allow rank-based segmentation to
##   be performed.
## 2014-05-20
## o Argument 'flavor' renamed to 'method'.
## o Argument 'statistic' renamed to 'stat'.
## o Removed arguments 'jitter', 'DP' and 'methModelSelection'. These
##   arguments may still be used through '...'.
## 2014-05-06
## o Changed the mapping: when all probes are not used : final
##   breakpoints are the median between two successive used probes and
##   not the used probes before the breakpoint.
## 2013-12-09
## o Replaced flavor 'cghseg' by 'DP'.
## 2013-12-06
## o Removed 'log(c)' statistic (left up to the user).
## o For flavors 'CnaStruct' and 'PSCN', force conversion of total CNs to log
##   scale if input data are not logged.
## 2013-11-29
## o Added flavor : 'DP'.
## o Cleanups in default arguments.
## 2013-03-28
## o Added flavors : 'CnaStruct' and 'Pelt'.
## 2013-03-07
## o Added option 'DP' for flavor "RBS" to do selection on initial segmentation.
## 2013-02-27
## o Bug fixed : flavor "GFLars" could not be run at full resolution.
## o Added statistic 'd|het'.
## 2013-02-26
## o Added option 'jitter' to allow more precise breakpoint identification by DP.
## 2013-02-15
## o Added flavor 'PSCBS'.
## 2013-01-28
## o Bug fix returned bkp from dp.
## 2013-01-25
## o Cleanups in doc and return values.
## 2013-01-09
## o Replaced all jumps by bkp.
## 2013-01-03
## o Added argument 'flavor' (passed to 'jointSeg').
## o Added flavors 'PSCN' and 'cghseg'.
## 2012-12-31
## o Added Argument 'statistic'.
## o Each dimension is now scaled to unit variance using 'estimateSd'.
## 2012-12-27
## o Some code and doc cleanups.
## 2012-12-15
## o Added argument 'profile' for optional reporting of CPU and memory usage.
## 2012-12-03
## o Added smoothing function from matrixStats.
## 2012-11-30
## o Created.
############################################################################

